pkg/controller/plan/adapter/ovirt/client.go

```diff
@@ -409,6 +418,107 @@ func (r *Client) cacert() []byte {
 	return nil
 }
 
-func (c *Client) Finalize(vms []*planapi.VMStatus, planName string) {
+func (r Client) Finalize(vms []*planapi.VMStatus, planName string) {
+	r.connect()
+	defer r.Close()
+	var wg sync.WaitGroup
+	wg.Add(len(vms))
+	for _, vm := range vms {
+		_, vmService, err := r.getVM(vm.Ref)
+		if err != nil {
+			return
+		}
+
+		go r.removePrecopies(vm.Warm.Precopies, vmService, &wg)
+	}
+
+	wg.Wait()
+}
+
+func (r Client) removePrecopies(precopies []planapi.Precopy, vmService *ovirtsdk.VmService, wg *sync.WaitGroup) {
+	if len(precopies) == 0 {
+		return
+	}
+
+	defer wg.Done()
+	snapsService := vmService.SnapshotsService()
+	for i := range precopies {
+		snapshotID := precopies[i].Snapshot
+		snapService := snapsService.SnapshotService(snapshotID)
+		correlationID := fmt.Sprintf("%s_finalize", snapshotID[0:8])
+		_, err := snapService.Remove().Query("correlation_id", correlationID).Send()
+		r.Log.Error(err, "Remove request for snapshot failed", "snapshotID", snapshotID)
+		if err != nil {
+			// Assume we could not remove because of an ongoing transfer
+			backoff := wait.Backoff{
+				Duration: time.Second * 3,
+				Factor:   0,
+				Jitter:   0,
+				Steps:    3,
+			}
```

Comment:
the description of the PR says this would apply to the first snapshot but this way we do it for all snapshots, no?

Reply:
yes, updated